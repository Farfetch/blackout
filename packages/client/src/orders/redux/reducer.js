/**
 * @module orders/reducer
 * @category Orders
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { adaptDate } from '../../helpers/adapters';
import { combineReducers } from 'redux';
import { reducerFactory } from '../../helpers/redux';
import get from 'lodash/get';
import merge from 'lodash/merge';
import omit from 'lodash/omit';
import produce from 'immer';

const INITIAL_STATE = {
  error: null,
  isLoading: false,
  result: null,
  ordersList: {
    error: null,
    isLoading: false,
  },
  orderDetails: {
    error: {},
    isLoading: {},
  },
  orderReturnOptions: {
    error: {},
    isLoading: {},
  },
  trackings: {
    error: null,
    isLoading: false,
  },
  documents: {
    error: null,
    isLoading: false,
  },
};

const error = (state = INITIAL_STATE.error, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_ORDER_DETAILS_FAILURE:
    case actionTypes.GET_ORDER_RETURN_OPTIONS_FAILURE:
    case actionTypes.GET_ORDERS_FAILURE:
    case actionTypes.GET_TRACKINGS_FAILURE:
      return action.payload.error;
    case actionTypes.GET_ORDER_DETAILS_REQUEST:
    case actionTypes.GET_ORDER_RETURN_OPTIONS_REQUEST:
    case actionTypes.GET_ORDERS_REQUEST:
    case actionTypes.GET_TRACKINGS_REQUEST:
    case actionTypes.RESET_ORDERS:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_ORDERS_REQUEST:
    case actionTypes.GET_ORDER_DETAILS_REQUEST:
    case actionTypes.GET_ORDER_RETURN_OPTIONS_REQUEST:
    case actionTypes.GET_TRACKINGS_REQUEST:
      return true;
    case actionTypes.GET_ORDERS_FAILURE:
    case actionTypes.GET_ORDERS_SUCCESS:
    case actionTypes.GET_ORDER_DETAILS_FAILURE:
    case actionTypes.GET_ORDER_DETAILS_SUCCESS:
    case actionTypes.GET_ORDER_RETURN_OPTIONS_FAILURE:
    case actionTypes.GET_ORDER_RETURN_OPTIONS_SUCCESS:
    case actionTypes.GET_TRACKINGS_FAILURE:
    case actionTypes.GET_TRACKINGS_SUCCESS:
    case actionTypes.RESET_ORDERS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const entitiesMapper = {
  [actionTypes.GET_ORDERS_SUCCESS]: (state, action) => {
    const orders = action.payload.entities.orders;
    const { entities } = action.payload;
    const merchantsFromOrders = {};

    for (const order in orders) {
      if (orders.hasOwnProperty(order)) {
        Object.values(orders[order].byMerchant).forEach(orderByMerchant => {
          merchantsFromOrders[orderByMerchant.merchant.id] =
            orderByMerchant.merchant;
          orderByMerchant.merchant = orderByMerchant.merchant.id;
        });
      }
    }

    // Reset orders before merging with entities
    const stateWithResetOrders = {
      ...state,
      orders: {},
    };

    return produce(stateWithResetOrders, draftState => {
      merge(draftState, entities);

      const newMerchants = merge(draftState.merchants, merchantsFromOrders);

      draftState.merchants = newMerchants;
    });
  },
  [actionTypes.GET_ORDER_DETAILS_SUCCESS]: (state, action) => {
    const { orderId } = action.meta;
    const { entities, result } = action.payload;
    // Filtering unimportant properties from the orderDetails request.
    const orderDetails = omit(result, [
      'baseUrl',
      'clientOnlyPage',
      'components',
      'countryCode',
      'countryId',
      'cultureCode',
      'currency',
      'currencyCode',
      'currencyCultureCode',
      'dataLayer',
      'isMobileDevice',
      'pageContent',
      'pageType',
      'redirectUrl',
      'relativeUrl',
      'requestSourceCountryCode',
      'returnUrl',
      'screenPixelsHeight',
      'screenPixelsWidth',
      'seoMetadata',
      'seoPageType',
      'serverSideJsApp',
      'slug',
      'staticPathorderDetails',
      'subfolder',
      'translationsUrl',
      'createdDate',
      'updatedDate',
    ]);
    const { createdDate, updatedDate } = result;
    const orderItems = entities.orderItems;

    const tempState = {
      ...state,
    };

    if (action.guest) {
      tempState.orders = {};
    }

    return produce(tempState, draftState => {
      merge(draftState, entities);
      for (const orderItem in orderItems) {
        if (orderItems.hasOwnProperty(orderItem)) {
          const merchantId = orderItems[orderItem].merchant;

          if (!draftState.orders || !draftState.orders[orderId]) {
            draftState.orders = {
              ...draftState.orders,
              [orderId]: { byMerchant: { [merchantId]: {} } },
            };
          }
          const orderByMerchant =
            draftState.orders[orderId].byMerchant[merchantId];
          const existingOrderItems = orderByMerchant?.orderItems || [];

          if (!existingOrderItems.includes(orderItem)) {
            draftState.orders[orderId].byMerchant[merchantId] = {
              ...orderByMerchant,
              orderItems: [...existingOrderItems, orderItem],
            };
          }
        }
      }

      draftState.orders[orderId] = {
        ...draftState.orders[orderId],
        ...orderDetails,
        createdDate: adaptDate(createdDate),
        updatedDate: adaptDate(updatedDate),
      };
    });
  },
  [actionTypes.GET_ORDER_RETURN_OPTIONS_SUCCESS]: (state, action) => {
    const { orderId } = action.meta;
    const returnOptions = action.payload.entities.returnOptions;
    const newReturnOptions = {};

    Object.values(returnOptions).forEach(value => {
      newReturnOptions[`${orderId}_${value.id}`] = {
        ...value,
        id: `${orderId}_${value.id}`,
      };
    });

    const newEntities = {
      ...action.payload.entities,
      returnOptions: newReturnOptions,
    };

    return produce(state, draftState => {
      merge(draftState, newEntities);
      for (const returnOption in newReturnOptions) {
        if (newReturnOptions.hasOwnProperty(returnOption)) {
          const merchantId = newReturnOptions[returnOption].merchant;

          const existingReturnOptions = get(
            draftState,
            `orders[${orderId}].byMerchant[${merchantId}].returnOptions`,
            [],
          );

          if (!existingReturnOptions.includes(returnOption)) {
            draftState.orders[orderId].byMerchant[merchantId] = {
              ...draftState.orders[orderId].byMerchant[merchantId],
              returnOptions: [...existingReturnOptions, returnOption],
            };
          }
        }
      }
    });
  },
  [actionTypes.RESET_ORDERS]: state => {
    const { orders, orderItems, ...remainingEntities } = state;

    return { ...remainingEntities };
  },
};

export const ordersList = reducerFactory(
  'GET_ORDERS',
  INITIAL_STATE.ordersList,
  actionTypes,
);

export const orderDetails = (
  state = INITIAL_STATE.orderDetails,
  action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_ORDER_DETAILS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.orderId]: true,
        },
        error: {
          ...state.error,
          [action.meta.orderId]: null,
        },
      };
    case actionTypes.GET_ORDER_DETAILS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.orderId]: false,
        },
      };
    case actionTypes.GET_ORDER_DETAILS_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.orderId]: false,
        },
        error: {
          ...state.error,
          [action.meta.orderId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const orderReturnOptions = (
  state = INITIAL_STATE.orderReturnOptions,
  action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_ORDER_RETURN_OPTIONS_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.meta.orderId]: true,
        },
        error: {
          ...state.error,
          [action.meta.orderId]: null,
        },
      };
    case actionTypes.GET_ORDER_RETURN_OPTIONS_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.orderId]: false,
        },
      };
    case actionTypes.GET_ORDER_RETURN_OPTIONS_FAILURE:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.meta.orderId]: false,
        },
        error: {
          ...state.error,
          [action.meta.orderId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const trackings = reducerFactory(
  'GET_TRACKINGS',
  INITIAL_STATE.trackings,
  actionTypes,
);

export const documents = reducerFactory(
  ['GET_ORDER_DOCUMENTS', 'GET_ORDER_DOCUMENT', 'POST_ORDER_DOCUMENT'],
  INITIAL_STATE.documents,
  actionTypes,
);

const result = (state = INITIAL_STATE.result, action = {}) => {
  switch (action.type) {
    case actionTypes.GET_ORDERS_SUCCESS:
      return action.payload.result;
    default:
      return state;
  }
};

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getResult = state => state.result;
export const getOrdersList = state => state.ordersList;
export const getOrderDetails = state => state.orderDetails;
export const getOrderReturnOptions = state => state.orderReturnOptions;
export const getTrackings = state => state.trackings;
export const getDocuments = state => state.documents;

/**
 * Reducer for orders state.
 *
 * @function ordersReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default combineReducers({
  error,
  isLoading,
  result,
  ordersList,
  orderDetails,
  orderReturnOptions,
  trackings,
  documents,
});
