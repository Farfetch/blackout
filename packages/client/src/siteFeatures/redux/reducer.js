// @TODO: Remove this file in version 2.0.0.
/**
 * @module siteFeatures/reducer
 * @category SiteFeatures
 * @subcategory Reducer
 */
import * as actionTypes from './actionTypes';

const INITIAL_STATE = {
  isLoading: false,
  error: null,
  result: null,
};

/**
 * Reducer for the siteFeatures state.
 *
 * @function siteFeaturesReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
const siteFeatures = (state = INITIAL_STATE, action = {}) => {
  switch (action.type) {
    case actionTypes.FETCH_SITE_FEATURES_REQUEST:
      return {
        ...state,
        isLoading: true,
        error: null,
      };
    case actionTypes.FETCH_SITE_FEATURES_SUCCESS:
      return {
        ...state,
        isLoading: false,
        error: null,
        result: action.payload.result,
      };
    case actionTypes.FETCH_SITE_FEATURES_FAILURE:
      return {
        ...state,
        isLoading: false,
        error: action.payload.error,
      };
    default:
      return state;
  }
};

export const getError = state => state.error;
export const getResult = state => state.result;
export const getIsLoading = state => state.isLoading;

export default siteFeatures;
