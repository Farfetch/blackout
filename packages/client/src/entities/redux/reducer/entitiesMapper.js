import { entitiesMapper as addressesEntitiesMapper } from '../../../addresses/redux';
import { entitiesMapper as authenticationEntitiesMapper } from '../../../authentication/redux';
import { entitiesMapper as bagEntitiesMapper } from '../../../../../redux/src/bags';
import { entitiesMapper as checkoutEntitiesMapper } from '../../../checkout/redux';
import { entitiesMapper as detailsEntitiesMapper } from '../../../products/details/redux';
import { entitiesMapper as listingEntitiesMapper } from '../../../products/listing/redux';
import { entitiesMapper as ordersEntitiesMapper } from '../../../orders/redux';
import { entitiesMapper as paymentsEntitiesMapper } from '../../../payments/redux';
import { entitiesMapper as profileEntitiesMapper } from '../../../profile/redux';
import { entitiesMapper as returnsEntitiesMapper } from '../../../returns/redux';
import { entitiesMapper as subscriptionsEntitiesMapper } from '../../../subscriptions/redux';
import createEntitiesReducer from './createEntities';

/**
 * Default entities mappers for all core funcionalities.
 */
export const defaultMappers = {
  addresses: addressesEntitiesMapper,
  authentication: authenticationEntitiesMapper,
  bag: bagEntitiesMapper,
  checkout: checkoutEntitiesMapper,
  details: detailsEntitiesMapper,
  listing: listingEntitiesMapper,
  orders: ordersEntitiesMapper,
  payments: paymentsEntitiesMapper,
  profile: profileEntitiesMapper,
  returns: returnsEntitiesMapper,
  subscriptions: subscriptionsEntitiesMapper,
};

export default ({ ...extraMappers }) =>
  createEntitiesReducer({
    ...addressesEntitiesMapper,
    ...authenticationEntitiesMapper,
    ...bagEntitiesMapper,
    ...checkoutEntitiesMapper,
    ...detailsEntitiesMapper,
    ...listingEntitiesMapper,
    ...ordersEntitiesMapper,
    ...paymentsEntitiesMapper,
    ...profileEntitiesMapper,
    ...returnsEntitiesMapper,
    ...subscriptionsEntitiesMapper,
    ...extraMappers,
  });
