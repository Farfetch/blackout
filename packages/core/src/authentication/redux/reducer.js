/**
 * @module authentication/reducer
 * @category Authentication
 * @subcategory Reducer
 */

import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { reducerFactory } from '../../helpers/redux';

const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  login: {
    error: null,
    isLoading: false,
  },
  logout: {
    error: null,
    isLoading: false,
  },
  register: {
    error: null,
    isLoading: false,
  },
  changePassword: {
    error: null,
    isLoading: false,
  },
  recoverPassword: {
    error: null,
    isLoading: false,
  },
  resetPassword: {
    error: null,
    isLoading: false,
  },
  validateEmail: {
    error: null,
    isLoading: false,
  },
  // TODO: Remove in version 2.0
  refreshToken: {
    error: null,
    isLoading: false,
  },
  refreshEmailToken: {
    error: null,
    isLoading: false,
  },
};

const error = (state = INITIAL_STATE.error, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_FAILURE:
    case actionTypes.LOGOUT_FAILURE:
    case actionTypes.REGISTER_FAILURE:
    case actionTypes.PASSWORD_CHANGE_FAILURE:
    case actionTypes.PASSWORD_RECOVER_FAILURE:
    case actionTypes.PASSWORD_RESET_FAILURE:
      return action.payload.error;
    case actionTypes.AUTHENTICATION_RESET:
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
    case actionTypes.REGISTER_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const id = (state = INITIAL_STATE.id, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
      return action.payload.result;
    case actionTypes.LOGOUT_SUCCESS:
    case actionTypes.AUTHENTICATION_RESET:
      return INITIAL_STATE.id;
    default:
      return state;
  }
};

const isLoading = (state = INITIAL_STATE.isLoading, action = {}) => {
  switch (action.type) {
    case actionTypes.LOGIN_REQUEST:
    case actionTypes.LOGOUT_REQUEST:
    case actionTypes.REGISTER_REQUEST:
    case actionTypes.PASSWORD_CHANGE_REQUEST:
    case actionTypes.PASSWORD_RECOVER_REQUEST:
    case actionTypes.PASSWORD_RESET_REQUEST:
      return true;
    case actionTypes.LOGIN_SUCCESS:
    case actionTypes.LOGOUT_SUCCESS:
    case actionTypes.REGISTER_SUCCESS:
    case actionTypes.PASSWORD_CHANGE_SUCCESS:
    case actionTypes.PASSWORD_RECOVER_SUCCESS:
    case actionTypes.PASSWORD_RESET_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

export const login = reducerFactory('LOGIN', INITIAL_STATE.login, actionTypes);
export const logout = reducerFactory(
  'LOGOUT',
  INITIAL_STATE.logout,
  actionTypes,
);
export const register = reducerFactory(
  'REGISTER',
  INITIAL_STATE.register,
  actionTypes,
);
export const changePassword = reducerFactory(
  'PASSWORD_CHANGE',
  INITIAL_STATE.changePassword,
  actionTypes,
);
export const resetPassword = reducerFactory(
  'PASSWORD_RESET',
  INITIAL_STATE.resetPassword,
  actionTypes,
);
export const recoverPassword = reducerFactory(
  'PASSWORD_RECOVER',
  INITIAL_STATE.recoverPassword,
  actionTypes,
);
export const validateEmail = reducerFactory(
  'VALIDATE_EMAIL',
  INITIAL_STATE.validateEmail,
  actionTypes,
);
export const refreshToken = reducerFactory(
  'REFRESH_TOKEN',
  INITIAL_STATE.refreshToken,
  actionTypes,
);
export const refreshEmailToken = reducerFactory(
  'REFRESH_EMAIL_TOKEN',
  INITIAL_STATE.refreshEmailToken,
  actionTypes,
);

export const entitiesMapper = {
  [actionTypes.LOGOUT_SUCCESS]: state => {
    return {
      ...state,
      user: {},
    };
  },
};

export const getError = state => state.error;
export const getIsLoading = state => state.isLoading;
export const getId = state => state.id;
export const getLogin = state => state.login;
export const getLogout = state => state.logout;
export const getRegister = state => state.register;
export const getChangePassword = state => state.changePassword;
export const getResetPassword = state => state.resetPassword;
export const getRecoverPassword = state => state.recoverPassword;
export const getValidateEmail = state => state.validateEmail;
export const getRefreshEmailToken = state => state.refreshEmailToken;
export const getRefreshToken = state => state.refreshToken; // @TODO: Remove in version 2.0

/**
 * Reducer for authentication state.
 *
 * @function authenticationReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default combineReducers({
  error,
  id,
  isLoading,
  login,
  logout,
  register,
  changePassword,
  resetPassword,
  recoverPassword,
  validateEmail,
  refreshToken, // @TODO: To remove in the next version 2.0
  refreshEmailToken,
});
