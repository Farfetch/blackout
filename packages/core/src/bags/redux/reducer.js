/**
 * @module bags/reducer
 * @category Bags
 * @subcategory Reducer
 */
import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createMergedObject } from '../../helpers/redux';

const INITIAL_STATE = {
  error: null,
  id: null,
  isLoading: false,
  bagItems: {
    error: {},
    isLoading: {},
  },
  bagOperations: {
    error: {},
    isLoading: {},
  },
};

const error = (
  state = INITIAL_STATE.error,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.ADD_ITEM_TO_BAG_FAILURE:
    case actionTypes.GET_BAG_FAILURE:
      return action.payload.error;
    case actionTypes.ADD_ITEM_TO_BAG_REQUEST:
    case actionTypes.DELETE_BAG_ITEM_REQUEST:
    case actionTypes.GET_BAG_REQUEST:
    case actionTypes.UPDATE_BAG_ITEM_REQUEST:
    case actionTypes.GET_BAG_OPERATION_REQUEST:
      return INITIAL_STATE.error;
    default:
      return state;
  }
};

const id = (
  state = INITIAL_STATE.id,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_BAG_SUCCESS:
      return action.payload.result;
    default:
      return state;
  }
};

const isLoading = (
  state = INITIAL_STATE.isLoading,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.ADD_ITEM_TO_BAG_REQUEST:
    case actionTypes.GET_BAG_REQUEST:
      return true;
    case actionTypes.ADD_ITEM_TO_BAG_FAILURE:
    case actionTypes.ADD_ITEM_TO_BAG_SUCCESS:
    case actionTypes.GET_BAG_FAILURE:
    case actionTypes.GET_BAG_SUCCESS:
      return INITIAL_STATE.isLoading;
    default:
      return state;
  }
};

const bagItems = (
  state = INITIAL_STATE.bagItems,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.DELETE_BAG_ITEM_REQUEST:
    case actionTypes.UPDATE_BAG_ITEM_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagItemId]: true,
        },
        error: {
          ...state.error,
          [action.payload.bagItemId]: null,
        },
      };
    case actionTypes.DELETE_BAG_ITEM_SUCCESS:
    case actionTypes.UPDATE_BAG_ITEM_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.bagItemId]: false,
        },
      };
    case actionTypes.DELETE_BAG_ITEM_FAILURE:
    case actionTypes.UPDATE_BAG_ITEM_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagItemId]: false,
        },
        error: {
          ...state.error,
          [action.payload.bagItemId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

const bagOperations = (
  state = INITIAL_STATE.bagOperations,
  /* istanbul ignore next */ action = {},
) => {
  switch (action.type) {
    case actionTypes.GET_BAG_OPERATION_REQUEST:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagOperationId]: true,
        },
        error: {
          ...state.error,
          [action.payload.bagOperationId]: null,
        },
      };
    case actionTypes.GET_BAG_OPERATION_SUCCESS:
      return {
        ...state,
        isLoading: {
          ...state.isLoading,
          [action.payload.bagOperationId]: false,
        },
      };
    case actionTypes.GET_BAG_OPERATION_FAILURE:
      return {
        isLoading: {
          ...state.isLoading,
          [action.payload.bagOperationId]: false,
        },
        error: {
          ...state.error,
          [action.payload.bagOperationId]: action.payload.error,
        },
      };
    default:
      return state;
  }
};

export const entitiesMapper = {
  [actionTypes.DELETE_BAG_ITEM_SUCCESS]: (
    state,
    /* istanbul ignore next */ action = {},
  ) => {
    const id = action.payload.result;
    const bag = action.payload.entities.bag[id];
    const newState = createMergedObject(state, action.payload.entities);

    newState.bag[id] = bag;

    return newState;
  },
  [actionTypes.RESET_BAG_ENTITIES]: state => {
    const { bag, bagItems, bagOperations, ...rest } = state;

    return rest;
  },
  [actionTypes.RESET_BAG_OPERATIONS]: state => {
    const { bagOperations, ...rest } = state;

    return rest;
  },
};

export const getError = state => state.error;
export const getId = state => state.id;
export const getIsLoading = state => state.isLoading;
export const getIsBagItemLoading = state => state.bagItems.isLoading;
export const getItemError = state => state.bagItems.error;
export const getIsBagOperationLoading = state => state.bagOperations.isLoading;
export const getBagOperationError = state => state.bagOperations.error;

const reducer = combineReducers({
  error,
  id,
  isLoading,
  bagItems,
  bagOperations,
});

/**
 * Reducer for bag state.
 *
 * @function bagReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
export default (state, action = {}) => {
  if (action.type === actionTypes.RESET_BAG_STATE) {
    const fieldsToReset = action.payload.fieldsToReset;

    if (!fieldsToReset) {
      return INITIAL_STATE;
    } else {
      const reducerFn = (acc, field) => {
        if (state.bagItems[field]) {
          return {
            ...acc,
            [field]: INITIAL_STATE[field],
            bagItems: {
              ...acc.bagItems,
              [field]: INITIAL_STATE.bagItems[field],
            },
          };
        }

        return { ...acc, [field]: INITIAL_STATE[field] };
      };

      return fieldsToReset.reduce(reducerFn, state);
    }
  }

  return reducer(state, action);
};
