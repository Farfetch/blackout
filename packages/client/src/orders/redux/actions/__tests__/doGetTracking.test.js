import * as normalizr from 'normalizr';
import {
  expectedTrackingNormalizedPayload,
  mockTrackingResponse,
} from '../../__fixtures__/orders.fixtures';
import { mockStore } from '../../../../../tests';
import doGetTracking from '../doGetTracking';
import find from 'lodash/find';
import reducer, { actionTypes } from '../../';
import thunk from 'redux-thunk';

const mockMiddlewares = [
  thunk.withExtraArgument({
    getOptions: () => ({ productImgQueryParam: '?c=2' }),
  }),
];
const ordersMockStore = (state = {}) =>
  mockStore({ orders: reducer() }, state, mockMiddlewares);

const normalizeSpy = jest.spyOn(normalizr, 'normalize');
const expectedConfig = undefined;
let store;

describe('doGetTracking() action creator', () => {
  const getTrackings = jest.fn();
  const action = doGetTracking(getTrackings);
  const trackingNumbers = '1';

  beforeEach(() => {
    jest.clearAllMocks();
    store = ordersMockStore();
  });

  it('should create the correct actions for when the get tracking procedure fails', async () => {
    const expectedError = new Error('get tracking error');

    getTrackings.mockRejectedValueOnce(expectedError);
    expect.assertions(4);

    try {
      await store.dispatch(action(trackingNumbers));
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(getTrackings).toHaveBeenCalledTimes(1);
      expect(getTrackings).toHaveBeenCalledWith(
        trackingNumbers,
        expectedConfig,
      );
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          { type: actionTypes.GET_TRACKINGS_REQUEST },
          {
            type: actionTypes.GET_TRACKINGS_FAILURE,
            payload: { error: expectedError },
          },
        ]),
      );
    }
  });

  it('should create the correct actions for when the get tracking procedure is successful', async () => {
    getTrackings.mockResolvedValueOnce(mockTrackingResponse);
    await store.dispatch(action(trackingNumbers));

    const actionResults = store.getActions();

    expect(normalizeSpy).toHaveBeenCalledTimes(1);
    expect(getTrackings).toHaveBeenCalledTimes(1);
    expect(getTrackings).toHaveBeenCalledWith(trackingNumbers, expectedConfig);

    expect(actionResults).toMatchObject([
      { type: actionTypes.GET_TRACKINGS_REQUEST },
      {
        type: actionTypes.GET_TRACKINGS_SUCCESS,
        payload: expectedTrackingNormalizedPayload,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.GET_TRACKINGS_SUCCESS,
      }),
    ).toMatchSnapshot('get tracking success payload');
  });
});
