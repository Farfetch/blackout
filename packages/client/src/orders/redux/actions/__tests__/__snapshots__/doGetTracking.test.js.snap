// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doGetTracking() action creator should create the correct actions for when the get tracking procedure is successful: get tracking success payload 1`] = `
Object {
  "payload": Object {
    "entities": Object {
      "courier": Object {
        "2": Object {
          "id": 2,
          "name": "DHL",
        },
      },
      "labelTracking": Object {
        "4538009162": Object {
          "courier": 2,
          "estimatedDeliveryDate": "2019-05-24T22:59:00Z",
          "events": Array [
            Object {
              "code": "PU",
              "date": "2019-05-23T18:35:23Z",
              "description": "Shipment picked up",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "PL",
              "date": "2019-05-23T19:37:46Z",
              "description": "Processed at DOCKLANDS-GBR",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "PL",
              "date": "2019-05-23T19:48:16Z",
              "description": "Processed at DOCKLANDS-GBR",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "DF",
              "date": "2019-05-23T19:48:58Z",
              "description": "Departed Facility in DOCKLANDS-GBR",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "AF",
              "date": "2019-05-23T20:39:47Z",
              "description": "Arrived at Sort Facility LONDON-HEATHROW-GBR",
              "location": "LONDON-HEATHROW-GBR",
              "signatory": "",
            },
            Object {
              "code": "PL",
              "date": "2019-05-23T22:25:12Z",
              "description": "Processed at LONDON-HEATHROW-GBR",
              "location": "LONDON-HEATHROW-GBR",
              "signatory": "",
            },
            Object {
              "code": "DF",
              "date": "2019-05-24T04:34:12Z",
              "description": "Departed Facility in LONDON-HEATHROW-GBR",
              "location": "LONDON-HEATHROW-GBR",
              "signatory": "",
            },
            Object {
              "code": "AR",
              "date": "2019-05-24T06:22:13Z",
              "description": "Arrived at Delivery Facility in CITY OF LONDON-GBR",
              "location": "CITY OF LONDON-GBR",
              "signatory": "",
            },
          ],
          "isEstimatedDeliveryDateTrusworthy": false,
          "service": "DOMESTIC EXPRESS",
          "trackingNumber": "4538009162",
        },
        "4538009163": Object {
          "courier": 2,
          "estimatedDeliveryDate": "2019-05-24T22:59:00Z",
          "events": Array [
            Object {
              "code": "PU",
              "date": "2019-05-23T18:35:23Z",
              "description": "Shipment picked up",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "PL",
              "date": "2019-05-23T19:37:46Z",
              "description": "Processed at DOCKLANDS-GBR",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "PL",
              "date": "2019-05-23T19:48:16Z",
              "description": "Processed at DOCKLANDS-GBR",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "DF",
              "date": "2019-05-23T19:48:58Z",
              "description": "Departed Facility in DOCKLANDS-GBR",
              "location": "DOCKLANDS-GBR",
              "signatory": "",
            },
            Object {
              "code": "AF",
              "date": "2019-05-23T20:39:47Z",
              "description": "Arrived at Sort Facility LONDON-HEATHROW-GBR",
              "location": "LONDON-HEATHROW-GBR",
              "signatory": "",
            },
            Object {
              "code": "PL",
              "date": "2019-05-23T22:25:12Z",
              "description": "Processed at LONDON-HEATHROW-GBR",
              "location": "LONDON-HEATHROW-GBR",
              "signatory": "",
            },
            Object {
              "code": "DF",
              "date": "2019-05-24T04:34:12Z",
              "description": "Departed Facility in LONDON-HEATHROW-GBR",
              "location": "LONDON-HEATHROW-GBR",
              "signatory": "",
            },
            Object {
              "code": "AR",
              "date": "2019-05-24T06:22:13Z",
              "description": "Arrived at Delivery Facility in CITY OF LONDON-GBR",
              "location": "CITY OF LONDON-GBR",
              "signatory": "",
            },
          ],
          "isEstimatedDeliveryDateTrusworthy": false,
          "service": "DOMESTIC EXPRESS",
          "trackingNumber": "4538009163",
        },
      },
    },
    "result": Object {
      "entries": Array [
        Object {
          "estimatedDeliveryDate": "2019-05-24T22:59:00Z",
          "events": Array [
            Object {
              "date": "2019-05-23T18:35:23Z",
              "description": "Pickup",
              "type": "Pickup",
            },
          ],
          "labelTrackings": Array [
            "4538009162",
          ],
          "numberOfTrackings": 1,
          "parcels": 2,
          "pickupDate": "2019-05-23T18:35:23Z",
        },
        Object {
          "estimatedDeliveryDate": "2019-05-24T22:59:00Z",
          "events": Array [
            Object {
              "date": "2019-05-23T18:35:23Z",
              "description": "Pickup",
              "type": "Pickup",
            },
          ],
          "labelTrackings": Array [
            "4538009163",
          ],
          "numberOfTrackings": 1,
          "parcels": 2,
          "pickupDate": "2019-05-23T18:35:23Z",
        },
      ],
      "number": 1,
      "totalItems": 2,
      "totalPages": 1,
    },
  },
  "type": "@farfetch/blackout-client/GET_TRACKINGS_SUCCESS",
}
`;
